dali/sasha/saarch.cpp

```diff
@@ -85,295 +85,444 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+                return false; //Log overflowed?
+            else if (outputModifiedTime)
+                cmd->addDT(modifiedTime);
+            return true;
+        }
+        bool addOutputFromWUTree(StringArray &outputFields, IPropertyTree *wuTree)
+        {
+            if (outputXML)
+            { //cmd->getAction()==SCA_GET
+                StringBuffer xml;
+                toXML(wuTree, xml);
+                if (!cmd->addResult(xml.str()))
+                    return false; //Log overflowed?
+            }
+            else
+            {
+                if (outputFields.length() == 0)
+                    cmd->addId(wuTree->queryName());
+                else
+                {
+                    StringBuffer output = wuTree->queryName();
+                    //Append more into the output.
+                    setWUDataTree(wuTree, outputFields, output);
+                    cmd->addId(output.str());
+                }
+            }
+            return true;
+        }
+        void setWUDataTree(IPropertyTree *wu, StringArray &outputFields, StringBuffer &output)
+        {
+            ForEachItemIn(i, outputFields)
+            {
+                const char* outputField = (const char*)outputFields.item(i);
+
+                StringBuffer val;
+                bool found = true;
+                if (strieq(outputField, "owner"))
+                    wu->getProp("@submitID", val);
+                else if (strieq(outputField, "cluster"))
+                    wu->getProp("@clusterName", val);
+                else if (strieq(outputField, "jobname"))
+                    wu->getProp("@jobName",val);
+                else if (strieq(outputField,"state"))
+                    wu->getProp("@state", val);
+                else
+                    found = false;
+                if (found)
+                    output.append(',').append(val);
+            }
+        }
+
+    public:
+
+        CArchivedWUReader(ISashaCommand *_cmd, const char *_mask)
+        {
+            cmd = _cmd;
+            mask.set(_mask);
+            isWild = mask.isEmpty() || isWildString(mask.str());
+
+            bool dfu = cmd->getDFU();
+            baseXPath.set(dfu ? "DFU/WorkUnits" : "WorkUnits");
+
+            //In WUiterate(), the queryAfter() is used as fromDT and the queryBefore() is used as toDT.
+            mkDateCompare(dfu, cmd->queryAfter(), fromDT, '0');
+            mkDateCompare(dfu, cmd->queryBefore(), toDT, '9');
+
+            beforeWU.set(cmd->queryBeforeWU());
+            afterWU.set(cmd->queryAfterWU());
+            checkBeforeOrAfterWU = !beforeWU.isEmpty() || !afterWU.isEmpty();
+            sortInc = !beforeWU.isEmpty() && afterWU.isEmpty();
+
+            numWUs = cmd->getLimit();
+            outputXML = cmd->getAction()==SCA_GET;
+            const char *outputFormat = cmd->queryOutputFormat();
+            outputFields.appendList(outputFormat, "|,");
+        }
+        void getWUs()
```

Comment:
This is dealing with archived workunits I think, would be better named getArchivedWUs

Reply:
It will be named to getArchivedWUs.